File: ScriptureAloneCore/Sources/ScriptureAloneCore/Import/ScriptureLabels.swift
// Importing is an iPhone, iPad and Mac feature: the watch has no file picker and no
// catalogue. It is also 32-bit (arm64_32), where the ZIP64 sentinel 0xFFFF_FFFF does not
// fit in an Int at all — so this code is not merely unused there, it cannot compile.
#if !os(watchOS)
import Foundation

/// Reading book names and references out of the strings publishers actually ship: chapter
/// headings ("The Gospel According to St. John", "PSALM 23") and element ids
/// ("ESV_Gen.1.1", "csb-Gen-1-1", "v12").
///
/// Book matching goes through the canon and abbreviation tables the app already has
/// (`Canon.swift`, `ReferenceParser`), so an ePub spelling the reader already understands
/// in the search bar is understood here too.
enum ScriptureLabels {
    /// What an element id turned out to name.
    enum Identifier: Sendable, Hashable {
        case verse(BookID, chapter: Int, verse: Int)
        case chapter(BookID, chapter: Int)
        /// `id="v3"` — a verse number with no book or chapter of its own.
        case verseNumber(Int)
        /// `id="ch2"` — a chapter number with no book of its own.
        case chapterNumber(Int)
    }

    /// Words that decorate a book title without naming it.
    private static let noise: Set<String> = [
        "the", "a", "book", "books", "gospel", "gospels", "according", "st", "saint", "holy",
        "bible", "epistle", "epistles", "letter", "letters", "general", "paul", "pauls", "apostle",
        "apostles", "jesus", "christ", "unto", "by", "chapter", "chapters",
    ]

    private static let verseWords: Set<String> = ["v", "vs", "ver", "vers", "verse", "versenum", "verseno"]
    private static let chapterWords: Set<String> = ["c", "ch", "chap", "chapter", "chapternum"]

    // MARK: - Headings

    /// A chapter heading: the book it names, the chapter number it carries, or both.
    /// Returns nil when the text is a section heading ("The Creation of the World") rather than a
    /// book or chapter title.
    static func heading(_ raw: String) -> (book: BookID?, chapter: Int?)? {
        var tokens = normalizedTokens(raw)
        guard !tokens.isEmpty, tokens.count <= 12 else { return nil }

        var chapter: Int?
        if let last = tokens.last, last.allSatisfy(\.isNumber), let value = Int(last), value > 0, value <= 200 {
            // A leading ordinal ("1 John") is part of the name, not a chapter number.
            if tokens.count > 1 || raw.trimmingCharacters(in: .whitespacesAndNewlines).allSatisfy({ $0.isNumber }) {
                chapter = value
                tokens.removeLast()
            }
        }
        guard let book = book(tokens: tokens) else {
            return chapter.map { (book: nil, chapter: $0) }
        }
        return (book: book, chapter: chapter)
    }

    /// The book a run of words names, trying the phrase as written and again with the connecting
    /// words removed, so both "Song of Solomon" and "First Epistle of John" resolve.
    ///
    /// The match must account for every word that is left after the decoration is stripped. That is
    /// what keeps a section heading like "Job’s Complaint" from being read as the book of Job.
    static func book(tokens: [String]) -> BookID? {
        let cleaned = tokens.filter { !noise.contains($0) }
        guard !cleaned.isEmpty else { return nil }
        let candidates = [cleaned, cleaned.filter { $0 != "of" && $0 != "to" }]
        var best: (book: BookID, width: Int)?
        for candidate in candidates {
            guard let match = longestMatch(candidate), match.width == candidate.count else { continue }
            if best == nil || match.width > best!.width { best = match }
        }
        if let best { return best.book }
        // Last resort: the app's own prefix matching, which knows every abbreviation the
        // search bar accepts. Kept narrow so prose headings do not become books.
        guard cleaned.count <= 2, let token = cleaned.last, token.count >= 3 else { return nil }
        let joined = ReferenceParser.normalizeOrdinals(cleaned.joined(separator: " "))
        return ReferenceParser.books(matching: joined).first
    }

    /// The longest run of adjacent words that is exactly a book name or abbreviation.
    private static func longestMatch(_ tokens: [String]) -> (book: BookID, width: Int)? {
        var best: (book: BookID, width: Int)?
        for start in tokens.indices {
            for end in stride(from: tokens.count, to: start, by: -1) {
                let width = end - start
                if let current = best, width <= current.width { break }
                let phrase = ReferenceParser.normalizeOrdinals(tokens[start..<end].joined(separator: " "))
                guard let book = exactBook(BookInfo.normalize(phrase)) else { continue }
                best = (book, width)
                break
            }
        }
        return best
    }

    /// Exact alias or USFM code — never a prefix, so "the" cannot become Titus.
    static func exactBook(_ token: String) -> BookID? {
        guard token.count >= 2 else { return nil }
        if let book = BookID.allCases.first(where: { $0.info.aliases.contains(token) }) { return book }
        return BookID.allCases.first { $0.code.lowercased() == token }
    }

    /// A book (and chapter) named by a file name: "gen01.xhtml", "01_Genesis.xhtml", "Genesis.xhtml".
    static func fileStem(_ stem: String) -> (book: BookID?, chapter: Int?)? {
        if let parsed = identifier(stem) {
            switch parsed {
            case .verse(let book, let chapter, _): return (book, chapter)
            case .chapter(let book, let chapter): return (book, chapter)
            case .chapterNumber(let chapter): return (nil, chapter)
            case .verseNumber: return nil
            }
        }
        var tokens = normalizedTokens(stem)
        // A zero-padded or large leading number is a sort key ("01_Genesis"); a bare 1, 2 or 3 is
        // an ordinal that belongs to the name ("1_John").
        while let first = tokens.first, first.allSatisfy(\.isNumber),
              first.count > 1 || (Int(first) ?? 0) > 3 {
            tokens.removeFirst()
        }
        guard !tokens.isEmpty else { return nil }
        return heading(tokens.joined(separator: " "))
    }

    static func normalizedTokens(_ raw: String) -> [String] {
        let lowered = raw.lowercased()
        var token = ""
        var tokens: [String] = []
        for character in lowered {
            if character.isLetter || character.isNumber {
                token.append(character)
            } else if !token.isEmpty {
                tokens.append(token)
                token = ""
            }
        }
        if !token.isEmpty { tokens.append(token) }
        return tokens
    }

    // MARK: - Identifiers

    /// Reads an element id. Handles `ESV_Gen.1.1`, `csb-Gen-1-1`, `MAT.5.3`, `1Cor.13.4`,
    /// `Gen.1`, `v12`, `verse-3`, `ch2`.
    static func identifier(_ raw: String) -> Identifier? {
        let parts = idComponents(raw)
        guard parts.count >= 2 else { return nil }

        func bookToken(at index: Int) -> BookID? {
            guard index >= 0, index < parts.count, !parts[index].isDigits else { return nil }
            let word = parts[index].text
            if index > 0, parts[index - 1].isDigits, parts[index - 1].text.count == 1,
               let ordinal = Int(parts[index - 1].text), (1...3).contains(ordinal),
               let book = exactBook(parts[index - 1].text + word) {
                return book
            }
            return exactBook(word)
        }

        let last = parts[parts.count - 1]
        guard last.isDigits, let lastValue = Int(last.text), lastValue > 0, lastValue < 1000 else { return nil }

        if parts.count >= 3, parts[parts.count - 2].isDigits,
           let chapter = Int(parts[parts.count - 2].text), chapter > 0, chapter < 1000,
           let book = bookToken(at: parts.count - 3) {
            return .verse(book, chapter: chapter, verse: lastValue)
        }
        if parts.count >= 2, !parts[parts.count - 2].isDigits {
            let word = parts[parts.count - 2].text
            if let book = bookToken(at: parts.count - 2) {
                return .chapter(book, chapter: lastValue)
            }
            if verseWords.contains(word) { return .verseNumber(lastValue) }
            if chapterWords.contains(word) { return .chapterNumber(lastValue) }
        }
        return nil
    }

    struct IDComponent: Sendable {
        var text: String
        var isDigits: Bool
    }

    /// Splits on separators and on letter/digit boundaries: "ESV_1Cor.13.4" -> ESV, 1, Cor, 13, 4.
    static func idComponents(_ raw: String) -> [IDComponent] {
        var parts: [IDComponent] = []
        var current = ""
        var currentIsDigits = false
        func flush() {
            if !current.isEmpty { parts.append(IDComponent(text: current, isDigits: currentIsDigits)) }
            current = ""
        }
        for character in raw.lowercased() {
            if character.isNumber {
                if !current.isEmpty && !currentIsDigits { flush() }
                currentIsDigits = true
                current.append(character)
            } else if character.isLetter {
                if !current.isEmpty && currentIsDigits { flush() }
                currentIsDigits = false
                current.append(character)
            } else {
                flush()
            }
        }
        flush()
        return parts
    }
}
#endif

File: android/app/src/main/java/com/blainemiller/scripturealone/data/importer/ScriptureLabels.kt
package com.blainemiller.scripturealone.data.importer

import com.blainemiller.scripturealone.data.canon.BookID
import com.blainemiller.scripturealone.data.reference.ReferenceParser

/**
 * Reading book names and references out of the strings publishers actually ship: chapter headings
 * ("The Gospel According to St. John", "PSALM 23") and element ids ("ABC_Gen.1.1", "xyz-Gen-1-1",
 * "v12"). Ported from `Import/ScriptureLabels.swift`.
 *
 * Book matching goes through the canon and abbreviation tables the app already has ([BookID],
 * [ReferenceParser]), so an ePub spelling the reader already understands in the search bar is
 * understood here too.
 */
internal object ScriptureLabels {
    /** What an element id turned out to name. */
    sealed class Identifier {
        data class Verse(val book: BookID, val chapter: Int, val verse: Int) : Identifier()
        data class Chapter(val book: BookID, val chapter: Int) : Identifier()

        /** `id="v3"` — a verse number with no book or chapter of its own. */
        data class VerseNumber(val verse: Int) : Identifier()

        /** `id="ch2"` — a chapter number with no book of its own. */
        data class ChapterNumber(val chapter: Int) : Identifier()
    }

    /** A place a heading or file name names: a book, a chapter number, or both. */
    data class Place(val book: BookID?, val chapter: Int?)

    /** Words that decorate a book title without naming it. */
    private val noise: Set<String> = setOf(
        "the", "a", "book", "books", "gospel", "gospels", "according", "st", "saint", "holy",
        "bible", "epistle", "epistles", "letter", "letters", "general", "paul", "pauls", "apostle",
        "apostles", "jesus", "christ", "unto", "by", "chapter", "chapters",
    )

    private val verseWords: Set<String> = setOf("v", "vs", "ver", "vers", "verse", "versenum", "verseno")
    private val chapterWords: Set<String> = setOf("c", "ch", "chap", "chapter", "chapternum")

    // MARK: - Headings

    /**
     * A chapter heading: the book it names, the chapter number it carries, or both. Returns null when
     * the text is a section heading ("The Creation of the World") rather than a book or chapter title.
     */
    fun heading(raw: String): Place? {
        val tokens = normalizedTokens(raw).toMutableList()
        if (tokens.isEmpty() || tokens.size > 12) return null

        var chapter: Int? = null
        val last = tokens.last()
        val value = if (SwiftText.allNumbers(last)) SwiftText.int(last) else null
        if (value != null && value > 0 && value <= 200) {
            // A leading ordinal ("1 John") is part of the name, not a chapter number.
            if (tokens.size > 1 || SwiftText.allNumbers(SwiftText.trimWhitespaceAndNewlines(raw))) {
                chapter = value
                tokens.removeAt(tokens.size - 1)
            }
        }
        val book = book(tokens) ?: return chapter?.let { Place(null, it) }
        return Place(book, chapter)
    }

    /**
     * The book a run of words names, trying the phrase as written and again with the connecting words
     * removed, so both "Song of Solomon" and "First Epistle of John" resolve.
     *
     * The match must account for every word that is left after the decoration is stripped. That is what
     * keeps a section heading like "Job’s Complaint" from being read as the book of Job.
     */
    fun book(tokens: List<String>): BookID? {
        val cleaned = tokens.filter { it !in noise }
        if (cleaned.isEmpty()) return null
        val candidates = listOf(cleaned, cleaned.filter { it != "of" && it != "to" })
        var best: Pair<BookID, Int>? = null
        for (candidate in candidates) {
            val match = longestMatch(candidate) ?: continue
            if (match.second != candidate.size) continue
            if (best == null || match.second > best.second) best = match
        }
        if (best != null) return best.first
        // Last resort: the app's own prefix matching, which knows every abbreviation the search bar
        // accepts. Kept narrow so prose headings do not become books.
        if (cleaned.size > 2 || SwiftText.characterCount(cleaned.last()) < 3) return null
        val joined = ReferenceParser.normalizeOrdinals(cleaned.joinToString(" "))
        return ReferenceParser.books(joined).firstOrNull()
    }

    /** The longest run of adjacent words that is exactly a book name or abbreviation. */
    private fun longestMatch(tokens: List<String>): Pair<BookID, Int>? {
        var best: Pair<BookID, Int>? = null
        for (start in tokens.indices) {
            var end = tokens.size
            while (end > start) {
                val width = end - start
                if (best != null && width <= best.second) break
                val phrase = ReferenceParser.normalizeOrdinals(tokens.subList(start, end).joinToString(" "))
                val book = exactBook(BookID.normalize(phrase))
                if (book != null) {
                    best = book to width
                    break
                }
                end--
            }
        }
        return best
    }

    /** Exact alias or USFM code — never a prefix, so "the" cannot become Titus. */
    fun exactBook(token: String): BookID? {
        if (SwiftText.characterCount(token) < 2) return null
        BookID.entries.firstOrNull { token in it.aliases }?.let { return it }
        return BookID.entries.firstOrNull { it.code.lowercase() == token }
    }

    /** A book (and chapter) named by a file name: "gen01.xhtml", "01_Genesis.xhtml", "Genesis.xhtml". */
    fun fileStem(stem: String): Place? {
        identifier(stem)?.let { parsed ->
            return when (parsed) {
                is Identifier.Verse -> Place(parsed.book, parsed.chapter)
                is Identifier.Chapter -> Place(parsed.book, parsed.chapter)
                is Identifier.ChapterNumber -> Place(null, parsed.chapter)
                is Identifier.VerseNumber -> null
            }
        }
        val tokens = normalizedTokens(stem).toMutableList()
        // A zero-padded or large leading number is a sort key ("01_Genesis"); a bare 1, 2 or 3 is an
        // ordinal that belongs to the name ("1_John").
        while (tokens.isNotEmpty()) {
            val first = tokens.first()
            if (!SwiftText.allNumbers(first)) break
            if (!(SwiftText.characterCount(first) > 1 || (SwiftText.int(first) ?: 0) > 3)) break
            tokens.removeAt(0)
        }
        if (tokens.isEmpty()) return null
        return heading(tokens.joinToString(" "))
    }

    fun normalizedTokens(raw: String): List<String> {
        val lowered = raw.lowercase()
        val characters = SwiftCharacters(lowered)
        val token = StringBuilder()
        val tokens = ArrayList<String>()
        for (i in 0 until characters.count) {
            if (characters.isLetter(i) || characters.isNumber(i)) {
                characters.appendTo(token, i)
            } else if (token.isNotEmpty()) {
                tokens.add(token.toString())
                token.setLength(0)
            }
        }
        if (token.isNotEmpty()) tokens.add(token.toString())
        return tokens
    }

    // MARK: - Identifiers

    /**
     * Reads an element id. Handles `ABC_Gen.1.1`, `xyz-Gen-1-1`, `MAT.5.3`, `1Cor.13.4`, `Gen.1`,
     * `v12`, `verse-3`, `ch2`.
     */
    fun identifier(raw: String): Identifier? {
        val parts = idComponents(raw)
        if (parts.size < 2) return null

        fun bookToken(index: Int): BookID? {
            if (index < 0 || index >= parts.size || parts[index].isDigits) return null
            val word = parts[index].text
            if (index > 0) {
                val previous = parts[index - 1]
                val ordinal = SwiftText.int(previous.text)
                if (previous.isDigits && SwiftText.characterCount(previous.text) == 1 && ordinal != null && ordinal in 1..3) {
                    exactBook(previous.text + word)?.let { return it }
                }
            }
            return exactBook(word)
        }

        val last = parts[parts.size - 1]
        val lastValue = SwiftText.int(last.text)
        if (!last.isDigits || lastValue == null || lastValue <= 0 || lastValue >= 1000) return null

        if (parts.size >= 3 && parts[parts.size - 2].isDigits) {
            val chapter = SwiftText.int(parts[parts.size - 2].text)
            if (chapter != null && chapter > 0 && chapter < 1000) {
                bookToken(parts.size - 3)?.let { return Identifier.Verse(it, chapter, lastValue) }
            }
        }
        if (parts.size >= 2 && !parts[parts.size - 2].isDigits) {
            val word = parts[parts.size - 2].text
            bookToken(parts.size - 2)?.let { return Identifier.Chapter(it, lastValue) }
            if (word in verseWords) return Identifier.VerseNumber(lastValue)
            if (word in chapterWords) return Identifier.ChapterNumber(lastValue)
        }
        return null
    }

    data class IDComponent(val text: String, val isDigits: Boolean)

    /** Splits on separators and on letter/digit boundaries: "ABC_1Cor.13.4" → ABC, 1, Cor, 13, 4. */
    fun idComponents(raw: String): List<IDComponent> {
        val parts = ArrayList<IDComponent>()
        val current = StringBuilder()
        var currentIsDigits = false
        fun flush() {
            if (current.isNotEmpty()) parts.add(IDComponent(current.toString(), currentIsDigits))
            current.setLength(0)
        }
        val characters = SwiftCharacters(raw.lowercase())
        for (i in 0 until characters.count) {
            if (characters.isNumber(i)) {
                if (current.isNotEmpty() && !currentIsDigits) flush()
                currentIsDigits = true
                characters.appendTo(current, i)
            } else if (characters.isLetter(i)) {
                if (current.isNotEmpty() && currentIsDigits) flush()
                currentIsDigits = false
                characters.appendTo(current, i)
            } else {
                flush()
            }
        }
        flush()
        return parts
    }
}
